Build integrated OpenCL Linux wheels

jgiannuzzi opened the pull request:
This PR builds on @tpboudreau's excellent work in #3144 to add support for building integrated OpenCL wheels on Linux too. It also changes the CI to build this wheel instead of the regular CPU wheel, offering a better out-of-the-box experience for Linux users who want to use LightGBM with GPU support.

Fixes #4684

jameslamb commented on tests/python_package_test/conftest.py:
```diff
@@ -0,0 +1,9 @@
+# `test_cpu_and_gpu_work` needs to be executed before any of the dask tests
+# because the AMD OpenCL driver won't load on Ubuntu 14.04 otherwise.
+# We simply make it first here.
```
Could you tell us more about this? What specifically happened and how did you narrow it down to `dask` tests being the problem?

If library-loading order affects whether or not this works, then I worry that `lightgbm`'s users might experience whatever issue you're working around with this change to the tests.

Linking a possibly-related conversation: https://github.com/microsoft/LightGBM/issues/5106#issuecomment-1125634723

jgiannuzzi replied:
The issue is that the statically-linked OpenCL ICD loader is not able to load the AMD App SDK driver that we use for testing on the CPU. The call to `dlopen` fails with the error `cannot load any more object with static TLS`. This seems to only happen with the AMD App SDK. I have tested with an NVIDIA GPU and with the PoCL CPU implementation and it worked fine with both.

When running under Python 3.8 on Ubuntu 14.04, the AMD App SDK driver fails to load immediately. With Python 3.7, it only fails to load if the `dask` tests run first. I isolated that by running various combinations and orders of tests.

The whole issue seems to come from the AMD App SDK driver using statically allocated thread local storage (STATIC_TLS), which means that the amount of TLS objects is limited, and the library loading order can influence when the limit gets reached.

It's worth noting that both neither the NVIDIA OpenCL driver, nor the PoCL one use STATIC_TLS, so the wheel works fine with them under various versions of Python. I do not have access to the required hardware, but I have checked that the various ICDs available on Ubuntu 20.04 (Mesa, Beignet, and Intel) do not use STATIC_TLS either.

jgiannuzzi replied:
I have tried running the tests with the `gpu` and `bdist` tasks with PoCL instead of the AMD App SDK and they all succeed, even the `dask` tests that are currently disabled with `gpu`. I think that using PoCL would make a lot of sense, as it is open source and actively maintained, as opposed to the AMD App SDK. It even enables testing of the Integrated OpenCL build on arm64!

I have a separate branch with these changes — if you're favourable to this idea, would you like me to raise another PR to get this done first, or would you like me to add it to this PR?

StrikerRUS replied:
@jgiannuzzi Thanks a lot for the detailed explanation!

> `cannot load any more object with static TLS`

Ah, I remember that annoying error! We've hit it already several times.
https://github.com/microsoft/LightGBM/pull/4893#issuecomment-1039204806
https://github.com/microsoft/LightGBM/pull/5022#issue-1146283008

> I think that using PoCL would make a lot of sense, as it is open source and actively maintained, as opposed to the AMD App SDK.

Absolutely agree with you!

> It even enables testing of the Integrated OpenCL build on arm64!

Awesome!

> I have a separate branch with these changes — if you're favourable to this idea, would you like me to raise another PR to get this done first, or would you like me to add it to this PR?

We'll really appreciate PR with those changes. I believe we can have it prior to this PR because due to low maintenance activity of this project we prefer small focused of one feature/fix self-contained PRs. Otherwise, big PRs can have a long review cycle.

jgiannuzzi replied:
Great, I opened #5282!

StrikerRUS replied:
We install AMD OpenCL for Linux here
https://github.com/microsoft/LightGBM/blob/11110c540eb84f519da299ea9eb76cc23a0ca570/.ci/setup.sh#L68-L82
and for Windows here
https://github.com/microsoft/LightGBM/blob/master/.ci/install_opencl.ps1
https://github.com/microsoft/LightGBM/blob/11110c540eb84f519da299ea9eb76cc23a0ca570/.ci/test_windows.ps1#L77-L78

StrikerRUS replied:
> Great, I opened https://github.com/microsoft/LightGBM/pull/5282!

Thanks a lot!

jameslamb reviewed (commented):
Thank you very much for working on this! 

I personally am not very familiar with OpenCL and not qualified to review this. Hopefully @shiyu1994 @guolinke and @StrikerRUS will be able to provide you some feedback. I'm also pinging @huanzhang12 to possibly help.

StrikerRUS commented:
@jgiannuzzi 
Thank you so much for working on this and for amazing PR!

> I personally am not very familiar with OpenCL and not qualified to review this.

Unfortunately, me too. But fortunately, I have an easy access to Ubuntu machine with NVIDIA GPU. So, I'll be able to independently test generated artifacts.

StrikerRUS commented on .vsts-ci.yml:
```diff
@@ -45,7 +45,8 @@ jobs:
       bdist:
         TASK: bdist
-        PYTHON_VERSION: '3.8'
+        # on Ubuntu 14.04, test_dual.py fails with newer version of Python
+        PYTHON_VERSION: '3.7'
```
Could you please share the exact error message?

jgiannuzzi replied:
This is the same error that we get when trying to do gpu-source with Python 3.8 on Ubuntu 14.04

StrikerRUS commented on CMakeLists.txt:
```diff
@@ -520,7 +520,7 @@ if(__INTEGRATE_OPENCL)
   target_include_directories(lightgbm_objs PRIVATE ${INTEGRATED_OPENCL_INCLUDES})
   target_compile_definitions(lightgbm_objs PRIVATE ${INTEGRATED_OPENCL_DEFINITIONS})
-  target_link_libraries(lightgbm_objs PUBLIC ${INTEGRATED_OPENCL_LIBRARIES})
+  target_link_libraries(lightgbm_objs PUBLIC ${INTEGRATED_OPENCL_LIBRARIES} ${CMAKE_DL_LIBS})
```
Could you please comment why this change was needed?

StrikerRUS commented:
@tpboudreau @itamarst We'll really appreciate your inputs for this PR!

StrikerRUS commented:
@jgiannuzzi Please update docs: https://github.com/microsoft/LightGBM/pull/3660/files#diff-e14b183376b1177323f6e7245d8ad64f2cd26f9638c129769d6b3c0ba4698dd5R26.

StrikerRUS commented on .ci/test.sh:
```diff
@@ -159,10 +159,12 @@ elif [[ $TASK == "bdist" ]]; then
+            # Make sure we can do both CPU and GPU; see tests/python_package_test/test_dual.py
+            export LIGHTGBM_TEST_DUAL_CPU_GPU=1
         else
             PLATFORM="manylinux2014_$ARCH"
```
>  It even enables testing of the Integrated OpenCL build on arm64!

Can we try to integrate OpenCL into `aarch64` Linux wheels?
https://github.com/microsoft/LightGBM/blob/ab9236baa06c3f071acf6b45a771c2a09661f146/.vsts-ci.yml#L158
Or maybe you've already tried to do it?

jgiannuzzi replied:
I have started to work on it. I had tested successfully with ubuntu 20.04, but then I noticed that we use CentOS in the CI. I hope to add a commit enabling aarch64 tomorrow.

StrikerRUS replied:
Sounds great!

jgiannuzzi replied:
Enabled in 6a08e4f and c4e0533

StrikerRUS replied:
Unfortunately, `test_dual.py` fails with `lightgbm.basic.LightGBMError: No OpenCL device found` on `aarch64`.
https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=12929&view=logs&j=c2f9361f-3c13-57db-3206-cee89820d5e3&t=bf5b33f9-9072-549e-d8b1-cab79a1dd2a9&l=977

StrikerRUS replied:
Maybe [`gpu_platform_id`](https://lightgbm.readthedocs.io/en/latest/Parameters.html#gpu_platform_id) and [`gpu_device_id`](https://lightgbm.readthedocs.io/en/latest/Parameters.html#gpu_device_id) params can help?..

jgiannuzzi replied:
This has been fixed in 0e25310. The issue was a wrong condition for detecting aarch64 (I was using the environment variable `ARCH` that was never set).

StrikerRUS commented on .ci/test.sh:
```diff
@@ -159,10 +159,12 @@ elif [[ $TASK == "bdist" ]]; then
             PLATFORM="manylinux2014_$ARCH"
         fi
-        cd $BUILD_DIRECTORY/python-package && python setup.py bdist_wheel --plat-name=$PLATFORM --python-tag py3 || exit -1
+        cd $BUILD_DIRECTORY/python-package && python setup.py bdist_wheel --integrated-opencl --plat-name=$PLATFORM --python-tag py3 || exit -1
```
Please update docs: https://github.com/microsoft/LightGBM/pull/3660/files#diff-e14b183376b1177323f6e7245d8ad64f2cd26f9638c129769d6b3c0ba4698dd5R26.

jgiannuzzi replied:
Will do, thanks for the reminder!

jgiannuzzi replied:
Done in 594a711

StrikerRUS replied:
Thanks a lot! But I think that `integrated-opencl` build option shouldn't be exposed to the public interface.
Could you please revert the following added sections?

![image](https://user-images.githubusercontent.com/25141164/174461389-26f14f2e-1823-438c-be18-aaf163b75269.png)

jgiannuzzi replied:
Done!

jgiannuzzi commented:
@StrikerRUS I have addressed your remaining comments.

@tpboudreau Could you please take a look? For the context, #5282 got merged before this one, switching the OpenCL CPU implementation to [PoCL](http://portablecl.org) on Linux.

StrikerRUS commented on tests/python_package_test/test_dual.py:
```diff
@@ -26,9 +27,10 @@ def test_cpu_and_gpu_work():
     params_gpu = params_cpu.copy()
     params_gpu["device"] = "gpu"
-    params_gpu["gpu_use_dp"] = True
+    params_gpu["gpu_use_dp"] = (platform.machine() == "x86_64")
```
Could you please explain why we don't set `gpu_use_dp=True` on aarch64?

jgiannuzzi replied:
The reason is that double-precision floats are not supported with PoCL on aarch64.

StrikerRUS replied:
Ah, I see! Thanks for the explanation.

Could you please add comment "double-precision floats are not supported with PoCL on aarch64" in the code to make it clear for everyone reading the code?

jgiannuzzi replied:
Done!

StrikerRUS commented:
We've recently moved PoCL installation from the CI runtime (`.ci/setup.sh`) to the Docker creation phase with the aim to reduce overall CI time: https://github.com/guolinke/lightgbm-ci-docker/pull/26 and #5286. That process introduced some merge conflicts in this PR. Sorry about that! Could you please resolve those conflicts?

StrikerRUS commented:
Good news!

I just checked generated as CI artifact wheel file on my Ubuntu machine with NVIDIA GPU. I simply installed it with `pip install *.whl` and after that `device='gpu'` parameter allowed me to utilize my GPU.

```python
>>> Python 3.7.12 | packaged by conda-forge | (default, Oct 26 2021, 06:08:53)
>>> [GCC 9.4.0] on linux

import numpy as np
import lightgbm as lgb

X = np.random.random((10_000, 200))
y = np.random.random(10_000)

est = lgb.LGBMRegressor(n_estimators=5000).fit(X, y)
>>> [LightGBM] [Warning] Auto-choosing col-wise multi-threading, the overhead of testing was 0.009667 seconds.
>>> You can set `force_col_wise=true` to remove the overhead.
>>> [LightGBM] [Info] Total Bins 51000
>>> [LightGBM] [Info] Number of data points in the train set: 10000, number of used features: 200
>>> [LightGBM] [Info] Start training from score 0.502931

est = lgb.LGBMRegressor(n_estimators=5000, device='gpu').fit(X, y)
>>> [LightGBM] [Info] This is the GPU trainer!!
>>> [LightGBM] [Info] Total Bins 51000
>>> [LightGBM] [Info] Number of data points in the train set: 10000, number of used features: 200
>>> [LightGBM] [Info] Using GPU Device: Tesla V100-SXM2-32GB, Vendor: NVIDIA Corporation
>>> [LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
>>> [LightGBM] [Info] GPU programs have been built
>>> [LightGBM] [Info] Size of histogram bin entry: 8
>>> [LightGBM] [Info] 200 dense feature groups (1.91 MB) transferred to GPU in 0.004207 secs. 0 sparse feature groups
>>> [LightGBM] [Info] Start training from score 0.502931
```

![image](https://user-images.githubusercontent.com/25141164/174499358-fad8e594-acb0-432e-b848-bf815685fffa.png)

StrikerRUS commented:
@jgiannuzzi Hey! Maybe we can merge this PR without the support for aarch64 and add it later in a separate PR?

jgiannuzzi commented:
Hey @StrikerRUS, I'm very sorry for not having updated this PR yet. I have had a fix for aarch64 for a while but never had the time to update the PR. I will try to do it this week so we can finally have those integrate OpenCL Linux wheels!

StrikerRUS commented:
@jgiannuzzi No problem. Thanks a lot for all your hard work!

ethanmdev commented:
So keen for this PR to get merged! @jgiannuzzi

jgiannuzzi commented:
Hey @StrikerRUS, I finally got around to updating this PR! I believe that I took into account all of your comments and I also fixed the support for aarch64.

It's taking 30 additional minutes because of building PoCL under emulation, which I believe we could solve in a separate PR by pre-building it (like we do for Ubuntu 14.04). I'm happy to take a shot at it later.

StrikerRUS commented:
> It's taking 30 additional minutes because of building PoCL under emulation, which I believe we could solve in a separate PR by pre-building it (like we do for Ubuntu 14.04). I'm happy to take a shot at it later.

That would be a little bit trickier because right now we use docker image from pypa org:
https://github.com/microsoft/LightGBM/blob/e063dad261d06049a461c3e2a364350e2b20b9f7/.vsts-ci.yml#L192
I think we can switch to using an our own one and referencing the current one as a base `FROM`.

StrikerRUS reviewed (approved):
Thank you so much for this PR!
All changes look good to me!

StrikerRUS commented:
> I think we can switch to using an our own one and referencing the current one as a base `FROM`.

Already started preparing a PR for this.
https://github.com/guolinke/lightgbm-ci-docker/compare/dev
Will continue after merging this PR.

jgiannuzzi commented:
Hi @jameslamb and @StrikerRUS, are you happy to merge this?

I'm looking forward to having nightly builds of LightGBM with GPU support on Linux, so let me know if there's anything else I can do!

StrikerRUS commented:
@jameslamb @jmoralez @guolinke @shiyu1994 Could you please help with a second review for this PR?

jameslamb commented:
I can review by the end of this week. Sorry for the delay @jgiannuzzi , and thanks for the awesome work!

shiyu1994 commented:
Thanks for working on this. This is super helpful. I'll come back to review this in this week, if by then no further review is provided.

jameslamb commented:
I'll be able to review today, sorry I didn't get to it yet!

jameslamb commented on .ci/setup.sh:
```diff
@@ -71,9 +71,40 @@ else  # Linux
+                gcc-c++ \
+                hwloc-devel \
+                sudo
+            yum install -y \
```
Why do these `yum install` blocks need to be split up? If there isn't a specific reason, could you please combine them into a single command (with alphabetized list of libraries)?

jgiannuzzi replied:
This is because installing `epel-release` adds a yum repository required for installing the packages in the second command. This can unfortunately not be done in a single command.

jameslamb replied:
ahhhh ok that makes sense. Thanks for the explanation.

jameslamb commented on .ci/setup.sh:
```diff
@@ -71,9 +71,40 @@ else  # Linux
+                llvm-toolset-7.0-clang-devel \
+                llvm-toolset-7.0-llvm-devel \
+                ocl-icd-devel
+            git clone --depth 1 --branch v1.8 https://github.com/pocl/pocl.git
```
In the time since this PR was first opened, PoCL has put up a new major release (v3.0, July 2022): https://github.com/pocl/pocl/releases.

Could you try upgrading to that, to take advantage of bugfixes and improvements in that project? I think that right now, while this support is new and we're actively testing it, is a great time to try to use the newest version possible.

Or, if that's a large amount of effort, after merging this would you please write up a new issue documenting the work to upgrade to that new version of PoCL?

jgiannuzzi replied:
Thanks for letting me know, I'll take a look at upgrading to PoCL v3.0!

For this specific PR, I have only aligned linux aarch64 to also use PoCL like we do on linux x86-64. I would thus suggest leaving the upgrade out of the scope of this PR.

jameslamb replied:
Ok nice, makes sense! Once we merge this PR, would you please document a feature request like "upgrade to PoCL v3.0" in this project's issue tracker? Just so we don't forget about. it.

jameslamb replied:
I created #5596 so we don't forget about this

jameslamb reviewed (changes requested):
Thanks very much for this! I left two suggestions for your consideration.

@StrikerRUS since we haven't been successful getting any other reviews with more OpenCL understanding to review...could you test the generated wheels one more time? There have been some additional commits since you last tested (https://github.com/microsoft/LightGBM/pull/5252#issuecomment-1159807098).

I'm also interested in understanding how much larger the package's wheels will be as a result of this change. Large wheels can introduce some new classes of problems for users (for example, those using function-as-a-service services that have fairly strict limits on disk usage, see https://github.com/pandas-dev/pandas/issues/30741#issuecomment-752751976). I'll wait until another commit is pushed to check that (since the previous Azure DevOps build is gone now).

jgiannuzzi commented:
> I'm also interested in understanding how much larger the package's wheels will be as a result of this change. Large wheels can introduce some new classes of problems for users (for example, those using function-as-a-service services that have fairly strict limits on disk usage, see [pandas-dev/pandas#30741 (comment)](https://github.com/pandas-dev/pandas/issues/30741#issuecomment-752751976)). I'll wait until another commit is pushed to check that (since the previous Azure DevOps build is gone now).

@jameslamb I have rebased my branch so that the CI can run again

jameslamb commented:
Thanks for that! I'll take a look shortly.

In the future, when contributing on this project please use merge commits instead of rebasing + force-pushing. Rebasing and force-pushing makes it more difficult for reviewers to see what was changed between review cycles, and breaks the link to our CI systems (which makes it much more difficult to trace back to "what happened in CI as of when this commit was pushed").

This project squashes all commits into 1 when merging, so it isn't necessary to keep a linear commit on PR branches.

jgiannuzzi commented:
Sorry about the rebase! I have reverted to the previous commit and merged `master` instead.

@jameslamb could you please take a look again once the CI has run?

jameslamb reviewed (commented):
Thanks very much for your patience and excellent contribution @jgiannuzzi !

We've finally fixed the CI issues in this project (#5510), and I'd like to help move this PR forward.

I don't have any other changes to recommend, but I would like to see CI run successfully and check the artifact size before approving. Can you please merge the latest changes from `master` into this branch?

I'm still hoping @StrikerRUS will be able to test again (https://github.com/microsoft/LightGBM/pull/5252#pullrequestreview-1113112842), but it's been a few weeks since I asked, so I think he may be unavailable for a while. If we don't hear from him in the next few days, I can spin up a box in AWS and test with something like the approach I took in https://github.com/microsoft/LightGBM/pull/5369#pullrequestreview-1064356007.

@guolinke @shiyu1994 could one of you help with a review as well?

jgiannuzzi commented:
> Can you please merge the latest changes from `master` into this branch?

@jameslamb done!

jgiannuzzi commented:
@jameslamb if it helps in any way, I have access to NVIDIA hardware (both for x86_64 and aarch64) and I'd be happy to help execute any tests you'd like to do. I can even do it with you through screen sharing if you'd like.

jgiannuzzi commented:
> I'm sorry, but I'm still struggling to get the artifacts from Azure DevOps. Either the retention for those artifacts is _very_ brief or I'm just doing something wrong.

I just checked and artifact publishing is not enabled for pull requests! I'm afraid you'll have to build the wheel locally to test it.

> Could you just tell me how large the wheels built this way are in your local setup? That'll give me a good enough idea to know if wheel size is something we need to optimize more for here.

I haven't done this in a while — I'll try to do this ASAP and report the size here.

jgiannuzzi commented:
@jameslamb the x86_64 package size with built-in OpenCL is 2.2MB.

Here is how I simulated a CI run to produce the wheel:

```bash
docker run --rm -ti \
--entrypoint bash \
-v $PWD/artifacts:/artifacts \
-e AZURE=true \
-e BUILD_DIRECTORY=/vsts/lightgbm \
-e BUILD_SOURCESDIRECTORY=/vsts/lightgbm \
-e BUILD_ARTIFACTSTAGINGDIRECTORY=/artifacts \
-e COMPILER=gcc \
-e CONDA_ENV=test-env \
-e LGB_VER=3.3.2.99 \
-e OS_NAME=linux \
-e PATH=/opt/miniforge/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin \
-e PRODUCES_ARTIFACTS=true \
-e PYTHON_VERSION=3.8 \
-e SETUP_CONDA=false \
-e TASK=bdist \
-e OMP_NUM_THREADS=4 \
-e POCL_MAX_PTHREAD_COUNT=4 \
lightgbm/vsts-agent:ubuntu-14.04 \
-c "git clone https://github.com/jgiannuzzi/lightgbm.git --recursive --depth 1 --branch linux-gpu-wheel && cd lightgbm && .ci/setup.sh && .ci/test.sh && chown -R $(id -u):$(id -g) /artifacts"
```
The last 2 environment variables are only needed on a machine with many cpu cores (mine has 256) to avoid some tests taking a huge amount of time. You can skip them if you have less than 16 cores (the CI does _not_ have them).

The wheel can be found under `artifacts` in the current working directory.

jgiannuzzi commented:
Hi @jameslamb @guolinke @shiyu1994! Is there anything I can do to help with getting this change merged?

jameslamb commented:
> the x86_64 package size with built-in OpenCL is 2.2MB

Amazing 🙌 

I was worried you were going to say like "125MB". 2.2MB compressed is great, that's the same order of magnitude as LightGBM's existing wheels (https://pypi.org/project/lightgbm/#files).

I'm sorry it's taken so long for me to test this out. We have been dealing with a serious lack of maintainer availability + a string of difficult CI issues (#5549, #5562, #5569).

Those CI issues are now resolved, and testing this is my top priority.

In the interim, can you please merge the latest changes from `master` into this branch?

Thank you so much for being patient with us. We really really appreciate this contribution and hope you'll continue contributing to LightGBM in the future.

jameslamb commented:
closing and re-opening to trigger CI

jameslamb commented:
I tried building this in CI last night (just pushing this branch from a fork to a LightGBM branch so it'd produce artifacts I could download and test with), and unfortunately I saw one failure on the `QEMU_multiarch bdist` build.

The `test_cpu_and_gpu_work()` test failed with the following.

> lightgbm.basic.LightGBMError: No OpenCL device found

([build link](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=13872&view=logs&j=c2f9361f-3c13-57db-3206-cee89820d5e3&t=bf5b33f9-9072-549e-d8b1-cab79a1dd2a9))

I just pushed https://github.com/microsoft/LightGBM/pull/5252/commits/a9f02d72db166eff8434deed5e08e729db885d09 updating this to latest `master`. Let's see if that happens here.

I'm so sorry @jgiannuzzi , I'm trying to test this but LightGBM's CI is not in a good state right now. I'm trying to fix it as fast as I can.

jameslamb commented:
I realized that in the new `manylinux_2_28_x86_64` image used to build `x86_64` linux wheels, a too-old OpenCL was being found, and as a result `pocl` didn't register itself in `/etc/OpenCL/vendors`.

Fixed that in https://github.com/guolinke/lightgbm-ci-docker/pull/29, which looks like it fixed https://github.com/microsoft/LightGBM/pull/5252#issuecomment-1321192047.

([ci build link](https://dev.azure.com/lightgbm-ci/lightgbm-ci/_build/results?buildId=13934&view=results))

I'll try fixing the other two CI failures tonight, they're almost certainly my fault. Sorry about that @jgiannuzzi , I'm doing my best to get this merged.

guolinke commented:
I am not sure why the  "guolinke/lightgbm-ci-docker#29"  will auto close this

jameslamb commented:
> I am not sure why the "https://github.com/guolinke/lightgbm-ci-docker/pull/29" will auto close this

oh strange! Probably in the way I phrased "fix" in that PR's description. Sorry about that and thanks for re-opening it.

jameslamb reviewed (approved):
Alright I've delayed this long enough, going to approve based on the passing CI and all I've learned about PoCL and this PR's other additions while working through #5580.

I'd still like to try these wheels again on different GPUs from a cloud provider, but that can be done later.

@guolinke can you look one more time? If you approve, please merge this (I'll be traveling for the next few days).

@jgiannuzzi thank you SO MUCH for this awesome contribution, and the other excellent contributions you've made to LightGBM along the way. I've learned a lot from you and it's been great working with you. I hope you'll consider contributing more to LightGBM in the future 😁

guolinke reviewed (approved):
Thank you!

jgiannuzzi commented:
Thank you @jameslamb and @guolinke, I'm looking forward to having daily builds with GPU support!

I'm sorry I wasn't available to help earlier with the many CI woes, and I'm certainly looking forward to contributing again to LightGBM in the future!
